Unable to install gevent on Windows 10 through cygwin

* gevent version: 1.5.0
* Python version: Please be as specific as possible: Python 3.6 from Cygwin installer
* Operating System: Windows 10 64-bit, Cygwin

### Description:

It appears that there are no prebuilt binaries for cygwin, so the installation attempts to build from source, however, it errors out during the link step.
Here are the errors along with the last gcc command that is run

> gcc -shared -Wl,--enable-auto-image-base build/temp.cygwin-3.1.4-x86_64-3.6/build/temp.cygwin-3.1.4-x86_64-3.6/gevent.libuv._corecffi.o build/temp.cygwin-3.1.4-x86_64-3.6/deps/libuv/src/fs-poll.o build/temp.cygwin-3.1.4-x86_64-3.6/deps/libuv/src/inet.o build/temp.cygwin-3.1.4-x86_64-3.6/deps/libuv/src/threadpool.o build/temp.cygwin-3.1.4-x86_64-3.6/deps/libuv/src/uv-common.o build/temp.cygwin-3.1.4-x86_64-3.6/deps/libuv/src/version.o build/temp.cygwin-3.1.4-x86_64-3.6/deps/libuv/src/uv-data-getter-setters.o build/temp.cygwin-3.1.4-x86_64-3.6/deps/libuv/src/timer.o build/temp.cygwin-3.1.4-x86_64-3.6/deps/libuv/src/idna.o build/temp.cygwin-3.1.4-x86_64-3.6/deps/libuv/src/strscpy.o build/temp.cygwin-3.1.4-x86_64-3.6/deps/libuv/src/unix/async.o build/temp.cygwin-3.1.4-x86_64-3.6/deps/libuv/src/unix/core.o build/temp.cygwin-3.1.4-x86_64-3.6/deps/libuv/src/unix/dl.o build/temp.cygwin-3.1.4-x86_64-3.6/deps/libuv/src/unix/fs.o build/temp.cygwin-3.1.4-x86_64-3.6/deps/libuv/src/unix/getaddrinfo.o build/temp.cygwin-3.1.4-x86_64-3.6/deps/libuv/src/unix/getnameinfo.o build/temp.cygwin-3.1.4-x86_64-3.6/deps/libuv/src/unix/loop-watcher.o build/temp.cygwin-3.1.4-x86_64-3.6/deps/libuv/src/unix/loop.o build/temp.cygwin-3.1.4-x86_64-3.6/deps/libuv/src/unix/pipe.o build/temp.cygwin-3.1.4-x86_64-3.6/deps/libuv/src/unix/poll.o build/temp.cygwin-3.1.4-x86_64-3.6/deps/libuv/src/unix/process.o build/temp.cygwin-3.1.4-x86_64-3.6/deps/libuv/src/unix/signal.o build/temp.cygwin-3.1.4-x86_64-3.6/deps/libuv/src/unix/stream.o build/temp.cygwin-3.1.4-x86_64-3.6/deps/libuv/src/unix/tcp.o build/temp.cygwin-3.1.4-x86_64-3.6/deps/libuv/src/unix/thread.o build/temp.cygwin-3.1.4-x86_64-3.6/deps/libuv/src/unix/tty.o build/temp.cygwin-3.1.4-x86_64-3.6/deps/libuv/src/unix/udp.o -L/usr/lib/python3.6/config -L/usr/lib -lpython3.6m -o build/lib.cygwin-3.1.4-x86_64-3.6/gevent/libuv/_corecffi.abi3.dll
  /usr/lib/gcc/x86_64-pc-cygwin/7.4.0/../../../../x86_64-pc-cygwin/bin/ld: build/temp.cygwin-3.1.4-x86_64-3.6/build/temp.cygwin-3.1.4-x86_64-3.6/gevent.libuv._corecffi.o: in function `_cffi_f_uv_fs_event_stop':
  /tmp/pip-install-2lkhphq6/gevent/build/temp.cygwin-3.1.4-x86_64-3.6/gevent.libuv._corecffi.c:2673: undefined reference to `uv_fs_event_stop'
  /tmp/pip-install-2lkhphq6/gevent/build/temp.cygwin-3.1.4-x86_64-3.6/gevent.libuv._corecffi.c:2673:(.text+0x5210): relocation truncated to fit: R_X86_64_PC32 against undefined symbol `uv_fs_event_stop'
  /usr/lib/gcc/x86_64-pc-cygwin/7.4.0/../../../../x86_64-pc-cygwin/bin/ld: build/temp.cygwin-3.1.4-x86_64-3.6/build/temp.cygwin-3.1.4-x86_64-3.6/gevent.libuv._corecffi.o: in function `_cffi_f_uv_fs_event_start':
  /tmp/pip-install-2lkhphq6/gevent/build/temp.cygwin-3.1.4-x86_64-3.6/gevent.libuv._corecffi.c:2635: undefined reference to `uv_fs_event_start'
  /tmp/pip-install-2lkhphq6/gevent/build/temp.cygwin-3.1.4-x86_64-3.6/gevent.libuv._corecffi.c:2635:(.text+0x538f): relocation truncated to fit: R_X86_64_PC32 against undefined symbol `uv_fs_event_start'
  /usr/lib/gcc/x86_64-pc-cygwin/7.4.0/../../../../x86_64-pc-cygwin/bin/ld: build/temp.cygwin-3.1.4-x86_64-3.6/build/temp.cygwin-3.1.4-x86_64-3.6/gevent.libuv._corecffi.o: in function `_cffi_f_uv_fs_event_init':
  /tmp/pip-install-2lkhphq6/gevent/build/temp.cygwin-3.1.4-x86_64-3.6/gevent.libuv._corecffi.c:2570: undefined reference to `uv_fs_event_init'
  /tmp/pip-install-2lkhphq6/gevent/build/temp.cygwin-3.1.4-x86_64-3.6/gevent.libuv._corecffi.c:2570:(.text+0x55af): relocation truncated to fit: R_X86_64_PC32 against undefined symbol `uv_fs_event_init'
  /usr/lib/gcc/x86_64-pc-cygwin/7.4.0/../../../../x86_64-pc-cygwin/bin/ld: build/temp.cygwin-3.1.4-x86_64-3.6/build/temp.cygwin-3.1.4-x86_64-3.6/gevent.libuv._corecffi.o: in function `_cffi_d_uv_fs_event_stop':
  /tmp/pip-install-2lkhphq6/gevent/build/temp.cygwin-3.1.4-x86_64-3.6/gevent.libuv._corecffi.c:2650: undefined reference to `uv_fs_event_stop'
  /tmp/pip-install-2lkhphq6/gevent/build/temp.cygwin-3.1.4-x86_64-3.6/gevent.libuv._corecffi.c:2650:(.text+0xd61): relocation truncated to fit: R_X86_64_PC32 against undefined symbol `uv_fs_event_stop'
  /usr/lib/gcc/x86_64-pc-cygwin/7.4.0/../../../../x86_64-pc-cygwin/bin/ld: build/temp.cygwin-3.1.4-x86_64-3.6/build/temp.cygwin-3.1.4-x86_64-3.6/gevent.libuv._corecffi.o: in function `_cffi_d_uv_fs_event_start':
  /tmp/pip-install-2lkhphq6/gevent/build/temp.cygwin-3.1.4-x86_64-3.6/gevent.libuv._corecffi.c:2585: undefined reference to `uv_fs_event_start'
  /tmp/pip-install-2lkhphq6/gevent/build/temp.cygwin-3.1.4-x86_64-3.6/gevent.libuv._corecffi.c:2585:(.text+0xd71): relocation truncated to fit: R_X86_64_PC32 against undefined symbol `uv_fs_event_start'
  /usr/lib/gcc/x86_64-pc-cygwin/7.4.0/../../../../x86_64-pc-cygwin/bin/ld: build/temp.cygwin-3.1.4-x86_64-3.6/build/temp.cygwin-3.1.4-x86_64-3.6/gevent.libuv._corecffi.o: in function `_cffi_d_uv_fs_event_init':
  /tmp/pip-install-2lkhphq6/gevent/build/temp.cygwin-3.1.4-x86_64-3.6/gevent.libuv._corecffi.c:2532: undefined reference to `uv_fs_event_init'
  /tmp/pip-install-2lkhphq6/gevent/build/temp.cygwin-3.1.4-x86_64-3.6/gevent.libuv._corecffi.c:2532:(.text+0xd81): relocation truncated to fit: R_X86_64_PC32 against undefined symbol `uv_fs_event_init'
  /usr/lib/gcc/x86_64-pc-cygwin/7.4.0/../../../../x86_64-pc-cygwin/bin/ld: build/temp.cygwin-3.1.4-x86_64-3.6/deps/libuv/src/unix/core.o: in function `uv_close':
  /tmp/pip-install-2lkhphq6/gevent/deps/libuv/src/unix/core.c:167: undefined reference to `uv__fs_event_close'
  /tmp/pip-install-2lkhphq6/gevent/deps/libuv/src/unix/core.c:167:(.text+0x91): relocation truncated to fit: R_X86_64_PC32 against undefined symbol `uv__fs_event_close'
  /usr/lib/gcc/x86_64-pc-cygwin/7.4.0/../../../../x86_64-pc-cygwin/bin/ld: build/temp.cygwin-3.1.4-x86_64-3.6/deps/libuv/src/unix/core.o: in function `uv__update_time':
  /tmp/pip-install-2lkhphq6/gevent/deps/libuv/src/unix/internal.h:307: undefined reference to `uv__hrtime'
  /tmp/pip-install-2lkhphq6/gevent/deps/libuv/src/unix/internal.h:307:(.text+0x388): relocation truncated to fit: R_X86_64_PC32 against undefined symbol `uv__hrtime'
  /usr/lib/gcc/x86_64-pc-cygwin/7.4.0/../../../../x86_64-pc-cygwin/bin/ld: build/temp.cygwin-3.1.4-x86_64-3.6/deps/libuv/src/unix/core.o: in function `uv_run':
  /tmp/pip-install-2lkhphq6/gevent/deps/libuv/src/unix/core.c:375: undefined reference to `uv__io_poll'
  /tmp/pip-install-2lkhphq6/gevent/deps/libuv/src/unix/core.c:375:(.text+0x434): relocation truncated to fit: R_X86_64_PC32 against undefined symbol `uv__io_poll'
  /usr/lib/gcc/x86_64-pc-cygwin/7.4.0/../../../../x86_64-pc-cygwin/bin/ld: build/temp.cygwin-3.1.4-x86_64-3.6/deps/libuv/src/unix/core.o: in function `uv__update_time':
  /tmp/pip-install-2lkhphq6/gevent/deps/libuv/src/unix/internal.h:307: undefined reference to `uv__hrtime'
  /tmp/pip-install-2lkhphq6/gevent/deps/libuv/src/unix/internal.h:307:(.text+0x5b6): relocation truncated to fit: R_X86_64_PC32 against undefined symbol `uv__hrtime'
  /usr/lib/gcc/x86_64-pc-cygwin/7.4.0/../../../../x86_64-pc-cygwin/bin/ld: /tmp/pip-install-2lkhphq6/gevent/deps/libuv/src/unix/internal.h:307: undefined reference to `uv__hrtime'
  /tmp/pip-install-2lkhphq6/gevent/deps/libuv/src/unix/internal.h:307:(.text+0x5ef): additional relocation overflows omitted from the output
  /usr/lib/gcc/x86_64-pc-cygwin/7.4.0/../../../../x86_64-pc-cygwin/bin/ld: /tmp/pip-install-2lkhphq6/gevent/deps/libuv/src/unix/internal.h:307: undefined reference to `uv__hrtime'
  /usr/lib/gcc/x86_64-pc-cygwin/7.4.0/../../../../x86_64-pc-cygwin/bin/ld: build/temp.cygwin-3.1.4-x86_64-3.6/deps/libuv/src/unix/core.o: in function `uv_hrtime':
  /tmp/pip-install-2lkhphq6/gevent/deps/libuv/src/unix/core.c:115: undefined reference to `uv__hrtime'
  /usr/lib/gcc/x86_64-pc-cygwin/7.4.0/../../../../x86_64-pc-cygwin/bin/ld: build/temp.cygwin-3.1.4-x86_64-3.6/deps/libuv/src/unix/core.o: in function `uv__io_close':
  /tmp/pip-install-2lkhphq6/gevent/deps/libuv/src/unix/core.c:950: undefined reference to `uv__platform_invalidate_fd'
  /usr/lib/gcc/x86_64-pc-cygwin/7.4.0/../../../../x86_64-pc-cygwin/bin/ld: build/temp.cygwin-3.1.4-x86_64-3.6/deps/libuv/src/unix/loop.o: in function `uv__update_time':
  /tmp/pip-install-2lkhphq6/gevent/deps/libuv/src/unix/internal.h:307: undefined reference to `uv__hrtime'
  /usr/lib/gcc/x86_64-pc-cygwin/7.4.0/../../../../x86_64-pc-cygwin/bin/ld: build/temp.cygwin-3.1.4-x86_64-3.6/deps/libuv/src/unix/loop.o: in function `uv_loop_init':
  /tmp/pip-install-2lkhphq6/gevent/deps/libuv/src/unix/loop.c:67: undefined reference to `uv__platform_loop_init'
  /usr/lib/gcc/x86_64-pc-cygwin/7.4.0/../../../../x86_64-pc-cygwin/bin/ld: /tmp/pip-install-2lkhphq6/gevent/deps/libuv/src/unix/loop.c:107: undefined reference to `uv__platform_loop_delete'
  /usr/lib/gcc/x86_64-pc-cygwin/7.4.0/../../../../x86_64-pc-cygwin/bin/ld: build/temp.cygwin-3.1.4-x86_64-3.6/deps/libuv/src/unix/loop.o: in function `uv_loop_fork':
  /tmp/pip-install-2lkhphq6/gevent/deps/libuv/src/unix/loop.c:118: undefined reference to `uv__io_fork'
  /usr/lib/gcc/x86_64-pc-cygwin/7.4.0/../../../../x86_64-pc-cygwin/bin/ld: build/temp.cygwin-3.1.4-x86_64-3.6/deps/libuv/src/unix/loop.o: in function `uv__loop_close':
  /tmp/pip-install-2lkhphq6/gevent/deps/libuv/src/unix/loop.c:148: undefined reference to `uv__platform_loop_delete'
  /usr/lib/gcc/x86_64-pc-cygwin/7.4.0/../../../../x86_64-pc-cygwin/bin/ld: build/temp.cygwin-3.1.4-x86_64-3.6/deps/libuv/src/unix/poll.o: in function `uv_poll_init':
  /tmp/pip-install-2lkhphq6/gevent/deps/libuv/src/unix/poll.c:74: undefined reference to `uv__io_check_fd'
  /usr/lib/gcc/x86_64-pc-cygwin/7.4.0/../../../../x86_64-pc-cygwin/bin/ld: build/temp.cygwin-3.1.4-x86_64-3.6/deps/libuv/src/unix/poll.o: in function `uv__poll_stop':
  /tmp/pip-install-2lkhphq6/gevent/deps/libuv/src/unix/poll.c:107: undefined reference to `uv__platform_invalidate_fd'
  /usr/lib/gcc/x86_64-pc-cygwin/7.4.0/../../../../x86_64-pc-cygwin/bin/ld: /tmp/pip-install-2lkhphq6/gevent/deps/libuv/src/unix/poll.c:107: undefined reference to `uv__platform_invalidate_fd'
  /usr/lib/gcc/x86_64-pc-cygwin/7.4.0/../../../../x86_64-pc-cygwin/bin/ld: build/temp.cygwin-3.1.4-x86_64-3.6/deps/libuv/src/unix/thread.o: in function `uv_cond_timedwait':
  /tmp/pip-install-2lkhphq6/gevent/deps/libuv/src/unix/thread.c:803: undefined reference to `uv__hrtime'
  collect2: error: ld returned 1 exit status
  error: command 'gcc' failed with exit status 1

The errors seems to be mentioning cffi, but i have that installed too

--- a/docs/changes/issue1645.feature
+++ b/docs/changes/issue1645.feature
@@ -1,0 +1,8 @@
+It is now possible to build and use the embedded libuv on a Cygwin
+platform.
+
+Note that Cygwin is not an officially supported platform of upstream
+libuv and is not tested by gevent, so the actual working status is
+unknown, and this may bitrot in future releases.
+
+Thanks to berkakinci for the patch.

--- a/src/gevent/libuv/_corecffi_build.py
+++ b/src/gevent/libuv/_corecffi_build.py
@@ -209,6 +209,20 @@ elif sys.platform.startswith('haiku'): # untested
     LIBUV_SOURCES += [
         _libuv_source('unix/haiku.c')
     ]
+elif sys.platform.startswith('cygwin'):
+    # Based on Cygwin package sources /usr/src/libuv-1.32.0-1.src/libuv-1.32.0/Makefile.am
+    # Apparently the same upstream at https://github.com/libuv/libuv/blob/v1.x/Makefile.am
+    LIBUV_SOURCES += [
+        _libuv_source('unix/cygwin.c'),
+        _libuv_source('unix/bsd-ifaddrs.c'),
+        _libuv_source('unix/no-fsevents.c'),
+        _libuv_source('unix/no-proctitle.c'),
+        _libuv_source('unix/posix-hrtime.c'),
+        _libuv_source('unix/posix-poll.c'),
+        _libuv_source('unix/procfs-exepath.c'),
+        _libuv_source('unix/sysinfo-loadavg.c'),
+        _libuv_source('unix/sysinfo-memory.c'),
+    ]
 
 
 LIBUV_MACROS = [

--- a/src/gevent/libuv/_corecffi_build.py
+++ b/src/gevent/libuv/_corecffi_build.py
@@ -171,7 +171,8 @@ elif sys.platform == 'darwin':
         _libuv_source('unix/kqueue.c'),
         _libuv_source('unix/proctitle.c'),
     ]
-elif sys.platform.startswith(('freebsd', 'dragonfly')):
+elif sys.platform.startswith(('freebsd', 'dragonfly')): # pragma: no cover
+    # Not tested
     LIBUV_SOURCES += [
         _libuv_source('unix/bsd-ifaddrs.c'),
         _libuv_source('unix/freebsd.c'),
@@ -179,7 +180,8 @@ elif sys.platform.startswith(('freebsd', 'dragonfly')):
         _libuv_source('unix/posix-hrtime.c'),
         _libuv_source('unix/bsd-proctitle.c'),
     ]
-elif sys.platform.startswith('openbsd'):
+elif sys.platform.startswith('openbsd'): # pragma: no cover
+    # Not tested
     LIBUV_SOURCES += [
         _libuv_source('unix/bsd-ifaddrs.c'),
         _libuv_source('unix/kqueue.c'),
@@ -187,7 +189,8 @@ elif sys.platform.startswith('openbsd'):
         _libuv_source('unix/posix-hrtime.c'),
         _libuv_source('unix/bsd-proctitle.c'),
     ]
-elif sys.platform.startswith('netbsd'):
+elif sys.platform.startswith('netbsd'): # pragma: no cover
+    # Not tested
     LIBUV_SOURCES += [
         _libuv_source('unix/bsd-ifaddrs.c'),
         _libuv_source('unix/kqueue.c'),
@@ -195,21 +198,26 @@ elif sys.platform.startswith('netbsd'):
         _libuv_source('unix/posix-hrtime.c'),
         _libuv_source('unix/bsd-proctitle.c'),
     ]
-elif sys.platform.startswith('sunos'):
+elif sys.platform.startswith('sunos'): # pragma: no cover
+    # Not tested.
     LIBUV_SOURCES += [
         _libuv_source('unix/no-proctitle.c'),
         _libuv_source('unix/sunos.c'),
     ]
-elif sys.platform.startswith('aix'):
+elif sys.platform.startswith('aix'): # pragma: no cover
+    # Not tested.
     LIBUV_SOURCES += [
         _libuv_source('unix/aix.c'),
         _libuv_source('unix/aix-common.c'),
     ]
-elif sys.platform.startswith('haiku'): # untested
+elif sys.platform.startswith('haiku'): # pragma: no cover
+    # Not tested
     LIBUV_SOURCES += [
         _libuv_source('unix/haiku.c')
     ]
-elif sys.platform.startswith('cygwin'):
+elif sys.platform.startswith('cygwin'): # pragma: no cover
+    # Not tested.
+
     # Based on Cygwin package sources /usr/src/libuv-1.32.0-1.src/libuv-1.32.0/Makefile.am
     # Apparently the same upstream at https://github.com/libuv/libuv/blob/v1.x/Makefile.am
     LIBUV_SOURCES += [
@@ -249,9 +257,9 @@ if sys.platform.startswith('linux'):
 elif sys.platform == 'darwin':
     _define_macro('_DARWIN_USE_64_BIT_INODE', 1)
     _define_macro('_DARWIN_UNLIMITED_SELECT', 1)
-elif sys.platform.startswith('netbsd'):
+elif sys.platform.startswith('netbsd'): # pragma: no cover
     _add_library('kvm')
-elif sys.platform.startswith('sunos'):
+elif sys.platform.startswith('sunos'): # pragma: no cover
     _define_macro('__EXTENSIONS__', 1)
     _define_macro('_XOPEN_SOURCE', 500)
     _add_library('kstat')
@@ -262,7 +270,7 @@ elif sys.platform.startswith('sunos'):
         # https://github.com/libuv/libuv/issues/1458
         # https://github.com/giampaolo/psutil/blob/4d6a086411c77b7909cce8f4f141bbdecfc0d354/setup.py#L298-L300
         _define_macro('SUNOS_NO_IFADDRS', '')
-elif sys.platform.startswith('aix'):
+elif sys.platform.startswith('aix'): # pragma: no cover
     _define_macro('_LINUX_SOURCE_COMPAT', 1)
     _add_library('perfstat')
 elif WIN:

